src/python/zquantum/core/interfaces/optimizer.py

```diff
@@ -105,3 +106,50 @@ def construct_history_info(cost_function, keep_history):
     if keep_history and hasattr(cost_function, "gradient"):
         histories["gradient_history"] = cost_function.gradient.history
     return histories
+
+
+class MetaOptimizer(ABC):
+    def __init__(
+        self,
+        ansatz: Ansatz,
+        inner_optimizer: Optimizer,
+        estimation_tasks_factory: Callable[[Any], EstimationTasksFactory],
+        cost_function_factory: Callable[[EstimationTasksFactory], CostFunction],
```

Comment:
I think that `cost_function_factory` should be able to take any inputs and it should be specified by particular MetaOptimizer implementation. Therefore the following signature might be better:
```suggestion
        cost_function_factory: Callable[[...], CostFunction],
```

However, I'm not sure if that's actually syntactically correct, @alexjuda could you take a look?